# STEP3_IN_SILICO_VALIDATION

## 1. Validation of our primers
Launch ecoPCR on a database formatted for the tool

This step is used to test the primers in silico. It allows us to check whether they correctly capture the desired targets and whether they also amplify other unexpected sequences.

### a. EcoPCR launch
***conda environment TaxonMarker_ecopcr***


To run an EcoPCR, use the script 1_commande_ecopcr.sh.

```bash=
./1_ecopcr_command.sh -h 

Usage: ./1_ecopcr_command.sh -a '<primer list>' -d '<path to .adx files>'
  -a   Comma-separated list of primer pairs, each consisting of a forward primer and a reverse primer. Primers are separated by a space and surrounded by an inverted comma. e.g. ‘TSRTCAAGAACRTBGARR SAYGTYCTGBACRTCRTC,TSRTCAAGAACRTBGARR AYGTYCTGNACRTCGTCV’.
  -d   Access path to the database for EcoPCR, the folder containing the .adx files. e.g. /BD_TaxonMarker/BD_ecoPCR/ecoPCR_db/assemblies/
```

This script then generates the EcoPCR commands. To run it:

**- If you are on a cluster**

Example:
```bash!
./1_ecopcr_command.sh -a 'GTNCCDCAYGGYGGYGG GCTTCNARDGCCCADACTT' -d $PATH_BD_ECOPCR_ASSEMBLIES/
```
Then run the command like this:
```bash=
sarray --mem=200G ecopcr_commands.sarray
```
**- If you are NOT on a cluster**
Comment on line 59 of the script 1_ecopcr_command.sh : 
```
# echo "module load devel/Miniconda/Miniconda3; source $MY_CONDA_PATH; conda activate TaxonMarker_EcoPCR; ecoPCR -d $(echo "$assembly" | sed 's/\.adx$//') $primer_f $primer_r > $ecopcr_outdir/$outdir/${name}.ecopcr;" >> $s_array_file
```
And uncomment line 61 of the 1_ecopcr_command.sh script:

```
echo "ecoPCR -d $(echo "$assembly" | sed 's/\.adx$//') $primer_f $primer_r > $ecopcr_outdir/$outdir/${name}.ecopcr;" >> $s_array_file
```
Launch 1_ecopcr_command.sh
```
./1_ecopcr_command.sh -a 'GTNCCDCAYGGYGGYGG GCTTCNARDGCCCADACTT' -d $PATH_BD_ECOPCR_ASSEMBLIES/
```

Enable your script to run 
```
 chmod +x ecopcr_commands.sarray
```
Activate the conda TaxonMarker_ecopcr environment and then run this command

```
./ecopcr_commands.sarray
```
### b. Taxonomic discrimination analysis

**Environment conda TaxonMarker_main**

#### 1_launch_format_ecopcr_result.sh

This script converts the ecoPCR results to FASTA format, adds the taxonomy information and then runs the vsearch tool to dereplicate the sequences. It also removes all sequences containing amino acids, which is necessary for SWARM to work properly.

We store the information about the sequences removed by dereplication in the output_vsearch_cluster.txt file. This allows us to add them to the SWARM results, so that we do not characterise a species as correctly discriminated when its equivalent has been deleted during dereplication.

Example:
```bash=
python ../script_treatment_ecopcr_result/format_ecopcr_result.py -o . -t $PATH_BD_ECOPCR/name_seq_with_taxo.txt result/1_GTNCCDCAYGGYGGYGG-GCTTCNARDGCCCADACTT/*ecopcr
```

The output fasta file is called **all_modified.fna**.

#### 2_launch_swarm.sh

We use clustering with the Swarm tool to measure species discrimination. Swarm is a highly discriminating, robust and fast tool that groups amplicons with strong similarities.

We assume that two amplicons grouped in the same cluster come from the same species.

Example:
```bash!
python ../script_treatment_ecopcr_result/Launch_swarm.py -f all_modified.fna -s fichier_swarm.txt -o cluster.txt  -t 4 -a 1 -d 1 -vsearch output_vsearch_cluster.txt -taxo $PATH_BD_ECOPCR/name_seq_with_taxo.txt
```

The output is :

- The raw file generated by Swarm (fichier_swarm.txt).
- A formatted file containing the Swarm results (cluster.txt), making the data more readable with the integration of all the taxonomies.

#### taxonomic_discrimination_at_species.py

This script is used to obtain measures of species discrimination, either specifically for a given taxonomic rank, or more generally. It allows you to include or exclude certain terms in the taxonomy in order to generate your own statistical file, limited only to the rank of interest.

Here are the parameters: 
```
  -h, --help            show this help message and exit
  -c CLUSTER_FILE, --cluster_file CLUSTER_FILE
                        Input file containing cluster information.
  -o OUTPUT_STATS_FILE, --output_stats_file OUTPUT_STATS_FILE
                        Output file for the statistics report.
  -r OUTPUT_CORRECTED_FILE, --output_corrected_file OUTPUT_CORRECTED_FILE
                        Output file for the filtered clusters.
  -i [INCLUDE_KEYWORDS ...], --include_keywords [INCLUDE_KEYWORDS ...]
                        Keywords that must be present in taxonomy to keep a line.
  -e [EXCLUDE_KEYWORDS ...], --exclude_keywords [EXCLUDE_KEYWORDS ...]
                        Keywords that, if present, will cause a line to be excluded.
  -l LOG_FILE, --log_file LOG_FILE
                        Output file to log rejected clusters (optional).
  --clean_words         Clean clusters based on a predefined dictionary of words.
  --all_species         If set, keeps the entire cluster if it contains at least one valid line.
  --selected_species    If set, removes clusters with only excluded lines but keeps clusters with mixed lines, removing only excluded lines.
```

**Additional explanation:**
- --i
This parameter is used to specify the taxonomic ranks of interest if you choose to filter with --all_species or --selected_species.

- --e 
This parameter is used to exclude words that are not yet present in the clean_words_dict dictionary (see below).

- --clean_words
The dictionary of words used to clean up clusters is :
```
clean_words_dict = {
    'sp.': True,
    'Unassigned': True,
    'unknown': True,
}
```

These terms indicate that the species are not yet well annotated. You can therefore choose to delete them in order to obtain only clean taxonomies.


**Cluster cleaning methods**
- --all_species

This option deletes clusters containing only non-interesting species.
However, if a cluster contains both species of interest and irrelevant species, it is retained.

This allows us to assess the discrimination of our species of interest against those captured unintentionally by the primers, without distorting the statistics with irrelevant species. In fact, we are not trying to discriminate species that are not of interest to us, so clusters containing only the latter are of no concern to us.

- --selected_species

This mode corresponds to a situation where we are working with a sample containing only the species of interest.
In this case, all irrelevant species are removed.

##### Example with our data_test

Let's take Lactobacillus as an example:

All members of this group no longer belong to the same genus as before, where they were all classified under the genus Lactobacillus. With the new classification, several genera have been created from the latter.

If I want my statistics file to represent Lactobacillus only, I'll run the following command:

```
python ../script_treatment_ecopcr_result/taxonomic_discrimination_at_species.py -c cluster.txt -o stats.txt -i f__Lactobacillaceae g__Acetilactobacillus g__Agrilactobacillus g__Amylolactobacillus g__Apilactobacillus g__Bombilactobacillus g__Companilactobacillus g__Fructilactobacillus g__Furfurilactobacillus g__Lacticaseibacillus g__Lactiplantibacillus g__Lactobacillus g__Lapidilactobacillus g__Latilactobacillus g__Lentilactobacillus g__Levilactobacillus g__Ligilactobacillus g__Limosilactobacillus g__Liquorilactobacillus g__Loigolactobacillus g__Paucilactobacillus g__Philodulcilactobacillus g__Paralactobacillus g__Schleiferilactobacillus g__Secundilactobacillus g__Xylocopilactobacillus g__Holzapfelia g__Dellaglioa -l rejected_clusters.txt -r cluster_corrected.txt --clean_words --selected_species
```

With the list of all the genera I want to keep in -i, along with the --clean_words and --selected_species parameters, I'll delete everything that isn't Lactobacillus.

I'll get :

- The file cluster_corrected.txt, which will only contain species belonging to the genera selected in their taxonomy.
- The rejected_clusters.txt file, which will contain all species not belonging to these genera.

The rejected_clusters.txt file will therefore include all species amplified by our primers that are not Lactobacillus.

NB: Specific inclusions and exclusions are not mandatory. We can choose to keep all results unchanged.

##### Explanation of script and results

Here's how the script works:

Note: We work on the complete taxonomy. If we only use the taxid, the results may vary, as two different taxids may have the same taxonomy if they correspond to strain taxids rather than species taxids. The script, however, uses a taxonomy that goes down to the species level.

The script assesses whether a taxonomy is “well discriminated” or “poorly discriminated” by examining the diversity of taxonomies present in each cluster.

For each cluster, the script extracts all the taxonomies associated with the sequences. If all sequences in a cluster belong to the same taxonomy (i.e. there is only one unique taxonomy in the cluster), the cluster is considered to have “good discrimination”. The taxonomy is then added to a set of “well-discriminated taxonomies”.

If a cluster contains several different taxonomies (i.e. several taxonomies are present in the same cluster), the cluster is considered to have “poor discrimination”. The script then adds the taxonomies present in this cluster to a set of “poorly discriminated taxonomies”.

The script uses sets to store good_discriminated_taxonomies and bad_discriminated_taxonomies.
Sets are data structures that do not allow duplicate elements to be stored. If a taxonomy is added to a set more than once, it will only be counted once.

Ambiguous taxonomies: If a taxonomy appears in both well-discriminated and poorly-discriminated clusters, it is considered “ambiguous” or having mixed discriminations.


Result for our previous order:

Résultat pour notre commande précédente:
```bash=
Number of unique taxonomies: 360 #Total number of different taxonomies in clusters
Number of well-discriminated taxonomies: 332
Percentage of well-discriminated taxonomies: 92.22%
Number of taxonomies both well and poorly discriminated: 47
Total number of filtered clusters: 633
Number of clusters with good discrimination: 582
Number of clusters with poor discrimination: 51
Percentage of clusters with good discrimination: 91.94%

# Below, we show poorly discriminated clusters and species with both good and poor discrimination.
```

## 2. Comparison with the V3V4 region of 16S rRNA (Example)

During the launch_format_ecopcr_result step, an output file named all_genome_name.txt is generated. It contains all the species identifiers amplified during our in silico PCR.

The idea is to assess whether the V3-V4 region of the 16S rRNA would have been more effective in discriminating these species.

```
 cd 2_Amplicon_comparison/
```
### a. Extraction of 16S rRNA sequences from the database
***Environment conda TaxonMarker_main***

Example:

```
python amplicon_sequence_selected_extractor.py --genome_file ../1_EcoPCR/all_genome_name.txt --tsv_file $PATH_BD_16S/16S_with_taxonomy.tsv --fasta_file $PATH_BD_16S/16S.fna --output_file filtered_16S.fna
```

### b. Formatting the fasta obtained with obiconvert for EcoPCR
***Environment conda TaxonMarker_ecopcr***

Example:
```
mkdir ecoPCR_db/
obiconvert --fasta filtered_16S.fna --ecopcrdb-output=ecoPCR_db/16S -t $PATH_NCBI_TAX_DUMB
```

### c. Launching pcr in silico
Example:
```
mkdir ecoPCR_db/
obiconvert --fasta filtered_16S.fna --ecopcrdb-output=ecoPCR_db/16S -t $PATH_NCBI_TAX_DUMB
```

### d. Analysis of results
***Environment conda TaxonMarker_main***
These are exactly the same scripts as for our primers.
Example:

```
python ../script_treatment_ecopcr_result/format_ecopcr_result.py -o . -t name_seq_amplicon.txt 16S.ecopcr


python ../script_treatment_ecopcr_result/Launch_swarm.py -f all_modified.fna -s fichier_swarm.txt -o cluster.txt  -t 4 -a 1 -d 1 -vsearch output_vsearch_cluster.txt -taxo name_seq_amplicon.txt


python ../script_treatment_ecopcr_result/taxonomic_discrimination_at_species.py -c cluster.txt -o stats.txt -i f__Lactobacillaceae g__Acetilactobacillus g__Agrilactobacillus g__Amylolactobacillus g__Apilactobacillus g__Bombilactobacillus g__Companilactobacillus g__Fructilactobacillus g__Furfurilactobacillus g__Lacticaseibacillus g__Lactiplantibacillus g__Lactobacillus g__Lapidilactobacillus g__Latilactobacillus g__Lentilactobacillus g__Levilactobacillus g__Ligilactobacillus g__Limosilactobacillus g__Liquorilactobacillus g__Loigolactobacillus g__Paucilactobacillus g__Philodulcilactobacillus g__Paralactobacillus g__Schleiferilactobacillus g__Secundilactobacillus g__Xylocopilactobacillus g__Holzapfelia g__Dellaglioa -l rejected_clusters.txt -r cluster_corrected.txt --clean_words --selected_species
```

Results:

```
Number of unique taxonomies: 340
Number of well-discriminated taxonomies: 143
Percentage of well-discriminated taxonomies: 42.06%
Number of taxonomies both well and poorly discriminated: 26
Total number of filtered clusters: 272
Number of clusters with good discrimination: 201
Number of clusters with poor discrimination: 71
Percentage of clusters with good discrimination: 73.90%
```
